// @angular/cli/models/command-runner.js
"use strict";
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:no-global-tslint-disable no-any
const core_1 = require("@angular-devkit/core");
const tools_1 = require("@angular-devkit/schematics/tools");
const fs_1 = require("fs");
const path_1 = require("path");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const yargsParser = require("yargs-parser");
const command_1 = require("../models/command");
const find_up_1 = require("../utilities/find-up");
const project_1 = require("../utilities/project");
const json_schema_1 = require("./json-schema");
// Based off https://en.wikipedia.org/wiki/Levenshtein_distance
// No optimization, really.
function levenshtein(a, b) {
    /* base case: empty strings */
    if (a.length == 0) {
        return b.length;
    }
    if (b.length == 0) {
        return a.length;
    }
    // Test if last characters of the strings match.
    const cost = a[a.length - 1] == b[b.length - 1] ? 0 : 1;
    /* return minimum of delete char from s, delete char from t, and delete char from both */
    return Math.min(levenshtein(a.slice(0, -1), b) + 1, levenshtein(a, b.slice(0, -1)) + 1, levenshtein(a.slice(0, -1), b.slice(0, -1)) + cost);
}
/**
 * Run a command.
 * @param args Raw unparsed arguments.
 * @param logger The logger to use.
 * @param context Execution context.
 */
function runCommand(args, logger, context, commandMap) {
    return __awaiter(this, void 0, void 0, function* () {
        // if not args supplied, just run the help command.
        if (!args || args.length === 0) {
            args = ['help'];
        }
        const rawOptions = yargsParser(args, { alias: { help: ['h'] }, boolean: ['help'] });
        let commandName = rawOptions._[0] || '';
        // remove the command name
        rawOptions._ = rawOptions._.slice(1);
        const executionScope = project_1.insideProject()
            ? command_1.CommandScope.inProject
            : command_1.CommandScope.outsideProject;
        if (commandMap === undefined) {
            const commandMapPath = find_up_1.findUp('commands.json', __dirname);
            if (commandMapPath === null) {
                logger.fatal('Unable to find command map.');
                return 1;
            }
            const cliDir = path_1.dirname(commandMapPath);
            const commandsText = fs_1.readFileSync(commandMapPath).toString('utf-8');
            const commandJson = JSON.parse(commandsText);
            commandMap = {};
            for (const commandName of Object.keys(commandJson)) {
                commandMap[commandName] = path_1.join(cliDir, commandJson[commandName]);
            }
        }
        let commandMetadata = commandName ? findCommand(commandMap, commandName) : null;
        if (!commandMetadata && (rawOptions.v || rawOptions.version)) {
            commandName = 'version';
            commandMetadata = findCommand(commandMap, commandName);
        }
        else if (!commandMetadata && rawOptions.help) {
            commandName = 'help';
            commandMetadata = findCommand(commandMap, commandName);
        }
        if (!commandMetadata) {
            if (!commandName) {
                logger.error(core_1.tags.stripIndent `
        We could not find a command from the arguments and the help command seems to be disabled.
        This is an issue with the CLI itself. If you see this comment, please report it and
        provide your repository.
      `);
                return 1;
            }
            else {
                const commandsDistance = {};
                const allCommands = Object.keys(commandMap).sort((a, b) => {
                    if (!(a in commandsDistance)) {
                        commandsDistance[a] = levenshtein(a, commandName);
                    }
                    if (!(b in commandsDistance)) {
                        commandsDistance[b] = levenshtein(b, commandName);
                    }
                    return commandsDistance[a] - commandsDistance[b];
                });
                logger.error(core_1.tags.stripIndent `
          The specified command ("${commandName}") is invalid. For a list of available options,
          run "ng help".
          Did you mean "${allCommands[0]}"?
      `);
                return 1;
            }
        }
        const command = yield createCommand(commandMetadata, context, logger);
        const metadataOptions = yield json_schema_1.convertSchemaToOptions(commandMetadata.text);
        if (command === null) {
            logger.error(core_1.tags.stripIndent `Command (${commandName}) failed to instantiate.`);
            return 1;
        }
        // Add the options from the metadata to the command object.
        command.addOptions(metadataOptions);
        let options = parseOptions(args, metadataOptions);
        args = yield command.initializeRaw(args);
        const optionsCopy = core_1.deepCopy(options);
        yield processRegistry(optionsCopy, commandMetadata);
        yield command.initialize(optionsCopy);
        // Reparse options after initializing the command.
        options = parseOptions(args, command.options);
        if (commandName === 'help') {
            options.commandInfo = getAllCommandInfo(commandMap);
        }
        if (options.help) {
            command.printHelp(commandName, commandMetadata.rawData.description, options);
            return;
        }
        else {
            const commandScope = mapCommandScope(commandMetadata.rawData.$scope);
            if (commandScope !== undefined && commandScope !== command_1.CommandScope.everywhere) {
                if (commandScope !== executionScope) {
                    let errorMessage;
                    if (commandScope === command_1.CommandScope.inProject) {
                        errorMessage = `This command can only be run inside of a CLI project.`;
                    }
                    else {
                        errorMessage = `This command can not be run inside of a CLI project.`;
                    }
                    logger.fatal(errorMessage);
                    return 1;
                }
                if (commandScope === command_1.CommandScope.inProject) {
                    if (!context.project.configFile) {
                        logger.fatal('Invalid project: missing workspace file.');
                        return 1;
                    }
                    if (['.angular-cli.json', 'angular-cli.json'].includes(context.project.configFile)) {
                        // --------------------------------------------------------------------------------
                        // If changing this message, please update the same message in
                        // `packages/@angular/cli/bin/ng-update-message.js`
                        const message = core_1.tags.stripIndent `
            The Angular CLI configuration format has been changed, and your existing configuration
            can be updated automatically by running the following command:

              ng update @angular/cli
          `;
                        logger.warn(message);
                        return 1;
                    }
                }
            }
        }
        delete options.h;
        delete options.help;
        yield processRegistry(options, commandMetadata);
        const isValid = yield command.validate(options);
        if (!isValid) {
            logger.fatal(`Validation error. Invalid command options.`);
            return 1;
        }
        return command.run(options);
    });
}
exports.runCommand = runCommand;
function processRegistry(options, commandMetadata) {
    return __awaiter(this, void 0, void 0, function* () {
        const rawArgs = options._;
        const registry = new core_1.schema.CoreSchemaRegistry([]);
        registry.addSmartDefaultProvider('argv', (schema) => {
            if ('index' in schema) {
                return rawArgs[Number(schema['index'])];
            }
            else {
                return rawArgs;
            }
        });
        const jsonSchema = json_schema_1.parseSchema(commandMetadata.text);
        if (jsonSchema === null) {
            throw new Error('');
        }
        yield registry.compile(jsonSchema).pipe(operators_1.concatMap(validator => validator(options)), operators_1.concatMap(validatorResult => {
            if (validatorResult.success) {
                return rxjs_1.of(options);
            }
            else {
                return rxjs_1.throwError(new core_1.schema.SchemaValidationException(validatorResult.errors));
            }
        })).toPromise();
    });
}
function parseOptions(args, optionsAndArguments) {
    const parser = yargsParser;
    // filter out arguments
    const options = optionsAndArguments
        .filter(opt => {
        let isOption = true;
        if (opt.$default !== undefined && opt.$default.$source === 'argv') {
            isOption = false;
        }
        return isOption;
    });
    const aliases = options
        .reduce((aliases, opt) => {
        if (!opt || !opt.aliases || opt.aliases.length === 0) {
            return aliases;
        }
        aliases[opt.name] = (opt.aliases || [])
            .filter(a => a.length === 1)[0];
        return aliases;
    }, {});
    const booleans = options
        .filter(o => o.type && o.type === 'boolean')
        .map(o => o.name);
    const defaults = options
        .filter(o => o.default !== undefined || booleans.indexOf(o.name) !== -1)
        .reduce((defaults, opt) => {
        defaults[opt.name] = opt.default;
        return defaults;
    }, {});
    const strings = options
        .filter(o => o.type === 'string')
        .map(o => o.name);
    const numbers = options
        .filter(o => o.type === 'number')
        .map(o => o.name);
    aliases.help = ['h'];
    booleans.push('help');
    const yargsOptions = {
        alias: aliases,
        boolean: booleans,
        default: defaults,
        string: strings,
        number: numbers,
    };
    const parsedOptions = parser(args, yargsOptions);
    // Remove aliases.
    options
        .reduce((allAliases, option) => {
        if (!option || !option.aliases || option.aliases.length === 0) {
            return allAliases;
        }
        return allAliases.concat([...option.aliases]);
    }, [])
        .forEach((alias) => {
        delete parsedOptions[alias];
    });
    // Remove undefined booleans
    booleans
        .filter(b => parsedOptions[b] === undefined)
        .map(b => core_1.strings.camelize(b))
        .forEach(b => delete parsedOptions[b]);
    // remove options with dashes.
    Object.keys(parsedOptions)
        .filter(key => key.indexOf('-') !== -1)
        .forEach(key => delete parsedOptions[key]);
    // remove the command name
    parsedOptions._ = parsedOptions._.slice(1);
    return parsedOptions;
}
exports.parseOptions = parseOptions;
// Find a command.
function findCommand(map, name) {
    // let Cmd: CommandConstructor = map[name];
    let commandName = name;
    if (!map[commandName]) {
        // find command via aliases
        commandName = Object.keys(map)
            .filter(key => {
            // get aliases for the key
            const metadataText = fs_1.readFileSync(map[key]).toString('utf-8');
            const metadata = JSON.parse(metadataText);
            const aliases = metadata['$aliases'];
            if (!aliases) {
                return false;
            }
            const foundAlias = aliases.filter((alias) => alias === name);
            return foundAlias.length > 0;
        })[0];
    }
    const metadataPath = map[commandName];
    if (!metadataPath) {
        return null;
    }
    const metadataText = fs_1.readFileSync(metadataPath).toString('utf-8');
    const metadata = core_1.parseJson(metadataText);
    return {
        path: metadataPath,
        text: metadataText,
        rawData: metadata,
    };
}
// Create an instance of a command.
function createCommand(metadata, context, logger) {
    return __awaiter(this, void 0, void 0, function* () {
        const schema = json_schema_1.parseSchema(metadata.text);
        if (schema === null) {
            return null;
        }
        const implPath = schema.$impl;
        if (typeof implPath !== 'string') {
            throw new Error('Implementation path is incorrect');
        }
        const implRef = new tools_1.ExportStringRef(implPath, path_1.dirname(metadata.path));
        const ctor = implRef.ref;
        return new ctor(context, logger);
    });
}
function mapCommandScope(scope) {
    let commandScope = command_1.CommandScope.everywhere;
    switch (scope) {
        case 'in':
            commandScope = command_1.CommandScope.inProject;
            break;
        case 'out':
            commandScope = command_1.CommandScope.outsideProject;
            break;
    }
    return commandScope;
}
function getAllCommandInfo(map) {
    return Object.keys(map)
        .map(name => {
        return {
            name: name,
            metadata: findCommand(map, name),
        };
    })
        .map(info => {
        if (info.metadata === null) {
            return null;
        }
        return {
            name: info.name,
            description: info.metadata.rawData.description,
            aliases: info.metadata.rawData.$aliases || [],
            hidden: info.metadata.rawData.$hidden || false,
        };
    })
        .filter(info => info !== null);
}
